File: cpp/10_classes.cpp
// ============================================================
// Topic:  Classes & Objects
// Summary: Define a class (state + behavior), construct instances, call methods.
// Run:     g++ -std=c++17 cpp/10_classes.cpp -o cpp/10_classes && cpp/10_classes
// ============================================================

#include <iostream>
#include <string>

// ----- 1. Define a class with a constructor and methods -----
class Person {
private:
    std::string name; // members are private by default in a class
    int age;

public:
    // ----- constructor (member initializer list) -----
    Person(const std::string& name, int age) : name(name), age(age) {}

    // ----- 2. Instance method -----
    std::string greet() const {
        return "Hi, I'm " + name + " and I'm " + std::to_string(age);
    }

    // ----- 3. Method that mutates instance state -----
    void birthday() { age += 1; }
};

int main() {
    // ----- 4. Create objects and call methods -----
    Person p1("Vetri", 22);          // stack object (no `new` needed)
    Person p2("Asha", 30);
    std::cout << p1.greet() << std::endl; // Output: Hi, I'm Vetri and I'm 22
    std::cout << p2.greet() << std::endl; // Output: Hi, I'm Asha and I'm 30

    p1.birthday();                        // mutate p1's state
    std::cout << p1.greet() << std::endl; // Output: Hi, I'm Vetri and I'm 23
    return 0;
}

// Cross-language note:
//   C++: members are private by default in `class` (public by default in `struct`).
//   Constructor has the class's name; `this` is a pointer. `const` methods can't mutate.
//   Objects can live on the stack (Person p;) — no `new` / garbage collector required.

File: typescript/10_classes.ts
// ============================================================
// Topic:  Classes & Objects
// Summary: Define a class (state + behavior), construct instances, call methods.
// Run:     cd typescript && npx ts-node 10_classes.ts
// ============================================================

// ----- 1. Define a class with a constructor and methods -----
class Person {
  name: string; // field declarations (required under strict mode)
  age: number;

  // constructor runs on `new`
  constructor(name: string, age: number) {
    this.name = name;
    this.age = age;
  }

  // ----- 2. Instance method (uses this) -----
  greet(): string {
    return `Hi, I'm ${this.name} and I'm ${this.age}`;
  }

  // ----- 3. Method that mutates instance state -----
  birthday(): void {
    this.age += 1;
  }
}

// ----- 4. Create objects (instances) with `new` and call methods -----
const p1 = new Person("Vetri", 22);
const p2 = new Person("Asha", 30);
console.log(p1.greet()); // Output: Hi, I'm Vetri and I'm 22
console.log(p2.greet()); // Output: Hi, I'm Asha and I'm 30

p1.birthday();           // mutate p1's state
console.log(p1.greet()); // Output: Hi, I'm Vetri and I'm 23

// Cross-language note:
//   TS: `constructor` keyword; instance is `this`. Use `new` to construct.
//   Shorthand: constructor(public name: string) auto-creates+assigns the field.
//   Fields are public by default; see 11_encapsulation for private/#.
